File: packages/zynth-keyboard/android/ZynthKeyboard/src/main/java/dev/zynth/keyboard/ZynthKeyboardBridge.kt
package dev.zynth.keyboard

import com.zynth.kit.runtime.ZynthModule
import com.zynth.kit.runtime.ZynthArgs
import org.json.JSONObject

/**
 * Bridge module that exposes keyboard functionality to JavaScript
 * via the __modules.call() mechanism.
 */
class ZynthKeyboardBridge(
    private val keyboardModule: ZynthKeyboardModule
) : ZynthModule {
    
    override val name: String = "ZynthKeyboard"

    override val exportedMethods: List<String> = listOf("dismiss", "getState", "setHeightSignalId")

    override val constants: Map<String, Any>?
        get() = keyboardModule.getInitialState().toMap()

    override fun call(method: String, args: ZynthArgs): JSONObject {
        return when (method) {
            "dismiss" -> resultResponse(handleDismiss())
            "getState" -> resultResponse(handleGetState())
            "setHeightSignalId" -> resultResponse(handleSetHeightSignalId(args))
            else -> throw IllegalArgumentException("Unsupported method: $method")
        }
    }

    private fun handleSetHeightSignalId(args: ZynthArgs): JSONObject {
        val id = args.getInt("id") ?: return JSONObject().put("success", false)
        keyboardModule.setHeightSignalId(id)
        return JSONObject().put("success", true)
    }

    private fun handleDismiss(): JSONObject {
        keyboardModule.dismissKeyboard()
        return JSONObject().put("success", true)
    }

    private fun handleGetState(): JSONObject {
        // Could be implemented to return current state if needed
        return JSONObject().put("success", true)
    }

    private fun resultResponse(result: Any?): JSONObject {
        return JSONObject().put("result", result)
    }
}

File: packages/zynth-keyboard/ios/ZynthKeyboardBridge.swift
//
//  ZynthKeyboardBridge.swift
//  ZynthKeyboard
//
//  Bridge module that exposes keyboard functionality to JavaScript
//  via the __modules.call() mechanism.
//

import Foundation
import ZynthKit

final class ZynthKeyboardBridge: NSObject, ZynthModule {

  let name = "ZynthKeyboard"
  private weak var keyboardModule: ZynthKeyboardModule?

  var exportedMethods: [String] {
    return ["dismiss", "getState"]
  }

  var constantsToExport: [String: Any]? {
    keyboardModule?.getInitialState().toDictionary()
  }

  init(keyboardModule: ZynthKeyboardModule) {
    self.keyboardModule = keyboardModule
    super.init()
  }

  func call(method: String, args: ZynthArgs) throws -> Any? {
    switch method {
    case "dismiss":
      return ["result": handleDismiss()]
    case "getState":
      return ["result": handleGetState()]
    default:
      throw ZynthModuleError.methodNotExported(module: name, method: method)
    }
  }

  private func handleDismiss() -> [String: Any] {
    keyboardModule?.dismissKeyboard()
    return ["success": true]
  }

  private func handleGetState() -> [String: Any] {
    return ["success": true]
  }
}
